### DISCUSSION TITLE
Save as Copy don't work as expected

### DISCUSSION THREAD

**Discussion by younky-yang** [directus/directus#13428] [Feature Requests]
Save as Copy function actually can save a lots of effort for user if they have item has many attributes and can be shared or just need to some subtle changes. But the current implementation of Save as Copy may not work as expected. 
1) when you create a new item, you can activate the Save as Copy function, but it doesn't work.
2) When you edit an item (without M2O or M2M relations), when you click Save as Copy, the edit information is gone, it just save the copy of original records. This should be the right one but from UI it should not be triggered after you made some changes. 
3) When you edit an item with M2O or M2M relations, the progress circle just spin for ever as it throws exception in SaveAsCopy().

I tried to analyze the source codes saveAsCopy() in app/src/composibles/use-item/use-item.ts, and have some thoughts on below. 
1) We give an popup menu options for item in list page, which user can just select Save As Copy without entering the detail page, with this approach, user will not be confused on the edit content lost, so it just save a copy of the selected item as new one
2) In the implementation of saveAsCopy(), we call an post API to just create a new copy of the selected item based on the PK, and handle all these relations in backend and then returns the new created copy item. 
3) Update the list page of collections with the new created copy item. 

### To Reproduce

create 2 collection and set some M2M relations, then edit one item in a collection, then trigger Save As Copy, you will get the progress circle spinning forever.

Update any collection item without M2M(M2O) collection, edit some content and then trigger Save As Copy, then you lost all the edits you made

### Errors Shown

The progress circle spinning forever without a message popup. or you just lost the edits you made for an item.

**Comment by azrikahar** [comment]
Understandably I did request you to open an issue over at https://github.com/directus/directus/pull/10418#issuecomment-1131319006, but as the points were either resolved, recently reported, or more discussive in nature, I'll move this in discussion for now.

> 1. when you create a new item, you can activate the Save as Copy function, but it doesn't work.

This was coincidentally reported very recently in #13407 and has been resolved 👍

> 2. When you edit an item (without M2O or M2M relations), when you click Save as Copy, the edit information is gone, it just save the copy of original records. This should be the right one but from UI it should not be triggered after you made some changes.

It should save the copy of the combined of original + edits, unless I'm misunderstanding here 🤔 Just to clarify, is this clip different from what you are experiencing if you were to do the same simple collection with name and description?

https://user-images.githubusercontent.com/42867097/169444947-38758199-aecb-4ece-86e7-7aee189a0d7a.mp4

> 3. When you edit an item with M2O or M2M relations, the progress circle just spin for ever as it throws exception in SaveAsCopy().

Also coincidentally same as #13406 since translations are also relational, so will update the title accordingly. (Actually assumed he was opening it in your stead 😄)

> I tried to analyze the source codes saveAsCopy() in app/src/composibles/use-item/use-item.ts, and have some thoughts on below.
>
> 1. We give an popup menu options for item in list page, which user can just select Save As Copy without entering the detail page, with this approach, user will not be confused on the edit content lost, so it just save a copy of the selected item as new one
>
> 2.In the implementation of saveAsCopy(), we call an post API to just create a new copy of the selected item based on the PK, and handle all these relations in backend and then returns the new created copy item.
>
> 3. Update the list page of collections with the new created copy item.

The popup menu approach is interesting! We will also need to see how that will work with the current `Duplication` > `Item Duplication Fields` configuration in collections, or possibly retire the latter for the former. Alternatively the latter can be the "default" selected fields perhaps. I'll defer to @benhaynes for his perspective here.

**Reply by younky-yang** [reply]
Thanks for this quick reply. 

Glad the first item was resolved so quickly. Will sync the code to have a test.
For the 2nd item, your screen video is the one I expected, but I had issue on this behavior because I have O2M relation, and the edits was missed, will try this later. 
https://github.com/directus/directus/issues/13406 This one is similar to my issue, and hope it can be fixed soon. 

Btw, there is another behaviors may not look right on Save As Copy. If an Item's PK is not auto generated, we should disable the function or promote with another user friendly managers, for now it shows PK is missing and user can't edit PK as it is disabled.

**Comment by ricricucit** [comment]
I've added a comment here, that might help someone debug: 
https://github.com/directus/directus/pull/10418#issuecomment-1368036653

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
